bam_sort.c

```diff
@@ -2636,6 +2849,8 @@ int bam_sort_core_ext(int is_by_qname, char* sort_by_tag, const char *fn, const
         free(fns);
     }
     bam_destroy1(b);
+    if (sam_order == TemplateCoordinate) {
```

Comment:
The number if items in use would be `k`.  You would need to free them after spilling to disk as well as at the end, or maybe check to see if the pointer has already been set so you can reuse them.  Although I'm not too enthusiastic with the idea of having a `malloc` per read as it potentially reintroduces the problems solved by PR #593, which were that `sort` didn't keep track of memory usage very well due to a lot of `malloc` overhead, and it spent a lot of time allocating and freeing small memory blocks.

You could get around this by putting `template_coordinate_key_t` directly in the `bam1_tag::u` union instead of using a pointer, but that would have the disadvantage of making the struct much bigger than it needs to be for sort orderings that don't need to look up the extra data.  I think better solutions might be one of:

1. Allocate `template_coordinate_key_t` structs in larger blocks, so you don't have so many of them to free.

2. Store the `template_coordinate_key_t` structs in `bam_mem` along with the BAM records.  This would have the advantage that they would be correctly accounted for in the memory usage.  You would need a way of dealing with the last record on each sorting round that fails to fit in the `bam_mem` block.

Option 1 is probably the easiest...